File: LogiOS/TrustarApp/Models/User/BinLocationTask.swift
//
//  BinLocationTask.swift
//  TrustarApp
//
//  Created by DionSoftware on 2022/02/23.
//  Copyright © 2022 DionSoftware. All rights reserved.
//

import Foundation
import CoreLocation
import Combine
import CoreMotion
import GRDB

class BinLocationTask {
    
    class BinState {
        var running: Bool
        var location: CLLocation? = nil
        
        init(running: Bool, location: CLLocation? = nil) {
            self.running = running
            self.location = location
        }
    }
    
    @Published private var workingBinHeaders =
    Current.Shared.userRepository!.binHeaderRepo.listPublisher()
        .map({ lst in
            return lst.filter { item in
                item.status.getStatus() == .Working
            }
            .map { it in
                return it.header
            }
        })
        .removeDuplicates()
        .eraseToAnyPublisher()
        .shareReplay(1)
    
    private var locationPublisher = PassthroughSubject<CLLocation, Never>()
    private var bag = Set<AnyCancellable>()
    private var userInfo: UserInfo
    
    init(userInfo: UserInfo) {
        self.userInfo = userInfo
        
        NotificationCenter.default.addObserver(self, selector: #selector(self.methodOfReceivedNotification(notification:)), name: Notification.Name(BroadcastEnum.Location.rawValue), object: nil)

        _ = $autoStart
            .subscribe(on: DispatchQueue.main)
            .scan(Optional<BinDetailAndStatus>(BinDetailAndStatus(detail: BinDetail(), status: WorkStatus()))) { [self] (acc, e) in
                let inRange = e?.filter({!($0.status.getStatus() == .Working || $0.inRange != true)})
                let b = acc != nil
                ? inRange?.first(where: { $0.binDetailAndStatus.sameItem(acc!) })
                : inRange?
                    .sorted(by: { getServiceOrder(binDetailAndStatus: $0.binDetailAndStatus) < getServiceOrder(binDetailAndStatus: $1.binDetailAndStatus)})
                    .sorted(by: { getWorkStatusOrder(wStatus: $0.status.getStatus()) < getWorkStatusOrder(wStatus: $1.status.getStatus()) })
                    .first

                return b?.binDetailAndStatus
            }
            .removeDuplicates(by: {b1, b2 -> Bool in
                let r = (b1 != nil && b2 != nil && b1!.sameItem(b2!)) || (b1 == nil && b2 == nil)
                return r
            })
            .map { data -> AnyPublisher<BinDetail?, Never> in
                let detail = data?.detail
                if detail == nil {
                    return Just(nil)
                        .eraseToAnyPublisher()
                        
                } else {
                    let delayInSecond = Double(BinDetail.stayTimeDelayinmin(binDetail: detail!)) * 60
                    return Just(detail)
                        .delay(for: .seconds(delayInSecond), scheduler: DispatchQueue.main)
                        .eraseToAnyPublisher()
                }
            }
            .switchToLatest()
            .subscribe(on: DispatchQueue.main)
            .retry(Int.max)
            .sink(receiveValue: { detail in
                let w = detail?.workCd
                if w != nil {
                    let workRepo = Current.Shared.userRepository?.workRepo
                    let _ = workRepo?.startWork(binDetail: detail!,
                                        work: Work(workCd: detail!.workCd!,
                                                   workNm: detail!.workNm ?? String(),
                                                   displayOrder: 0,
                                                   displayFlag: false),
                                        fromLocation: Current.Shared.lastLocation,
                                        addIfFinished: true,
                                        isAutoMode: true)
                    Helper.Shared.sendBroadcastNotification(.Bin)
                    Current.Shared.syncBin() {(_,_) in }
                }
            })
        
        let user = self.userInfo
        //休憩判定
        restEvent(cancelableBags: &bag,
                  working: workingBinHeaders,
                  location: locationPublisher.eraseToAnyPublisher(),
                  user: user)
            .sink { (allocationNo, r) in
                if (r is DetectRestEvent.MoveDetected) {
                    let r2 = (r as! DetectRestEvent.MoveDetected)
                    let c = CommonRest(userInfor: user,
                                       allocationNo: allocationNo,
                                       startLocation: r2.rest.startRestLocation,
                                       endDate: r2.restEndTimeBasedOsDate(),
                                       elapsed: r2.elapsedRestMillis())
                    do {
                        try Config.Shared.resultDb?.instanceDb?.write { db in
                            try c.insert(db)
                        }
                    } catch {
                        debugPrint(error)
                    }
                    Current.Shared.syncRest() {(_,_) in }
                }
            }
            .store(in: &bag)
        
        let u = self.userInfo
        let d = u.decimationRange
        let detectStartTimingInMils = Int64(userInfo.detectStartTiming)! * 60_000
        let dao = Config.Shared.userDb?.sensorDetectEventDao
        
        //センサー検知
        sensorDetect(cancelableBags: &bag,
                     working: workingBinHeaders,
                     binHeaderDelayInMillis: detectStartTimingInMils,
                     workDelayInMillis: 5000,
                     sensorManager: sensorManager,
                     detectControl: Current.Shared.userRepository!.binDetailRepo.countWorkingStatus
        ).sink(receiveValue: { item in
            if item is SensorRecord.DropRecent {
                let item2 = (item as! SensorRecord.DropRecent)
                let no = item2.allocationNo
                dao!.deleteByTimeRange(allocationNo: no, from: item2.from, to: item2.to)
                if item2.isFinished {
                    let ls = dao!.findBy(allocationNo: no)
                    if !ls.isEmpty {
                        let dt = ls.map { $0.csvRow(formatter: Config.Shared.dateFormatterSensorCsv)}
                            .joined(separator: "\n")
                        
                        let csv = CommonSensorCsv(userInfor: userInfo, allocationNo: no, csv: Data(dt.utf8))
                        
                        do {
                            try Config.Shared.resultDb?.instanceDb?.write { db in
                                try csv.insert(db)
                            }
                            
                            dao?.deleteBy(allocationNo: no)
                            
                        } catch {
                            debugPrint(error)
                        }
                    }
                }
            }
            if item is SensorRecord.Record {
                let item2 = (item as! SensorRecord.Record)
                let n = item2.event
                let l = dao?.latestEvent(allocationNo: n.allocationNo)?.distanceTo(location: CLLocation(latitude: n.eventRecordLatitude!, longitude: n.eventRecordLongitude!))
                if l == nil || d == nil || l! > Double(d!) {
                    do {
                        try Config.Shared.userDb?.instanceDb?.write { db in
                            try n.insert(db)
                        }
                    } catch {
                        debugPrint(error)
                    }
                    
                }
            }
        })
        .store(in: &bag)

        //センサー検知レコード位置更新
        locationPublisher.sink { location in
            dao?.latestEventEachGroup()
                .map({ sde -> SensorDetectEvent? in
                    let prevLocationTime: Int64 = sde.locationTimestamp ?? 0
                    let currLocationTime = location.timestamp.milisecondsSince1970
                    let eventTime = Int64(sde.eventRecordDate)
                    let td1 = abs(eventTime - prevLocationTime)
                    let td2 = abs(eventTime - currLocationTime)
                    if td1 > td2 {
                        return sde.copyOf(newLocation: location)
                    } else {
                        return nil
                    }
                })
                .compactMap{$0}
                .forEach({ it in
                    do {
                        try Config.Shared.userDb?.instanceDb?.write { db in
                            try it.save(db)
                        }
                    } catch {
                        debugPrint(error)
                    }
                })
        }
        .store(in: &bag)

        let skipRange = user.decimationRange ?? 50
        let skipInterval = user.positionGetInterval
        
        //位置レコード
        workingBinHeaders
            .eachWorkingBin(cancelableBags: &bag) { b in
            return self.locationPublisher.filterWithLast { upStreamLast, downStreamLast, current in
                !(shouldSkipLocation(range: skipRange, current: current, previous: upStreamLast, lastRecorded: downStreamLast) ||
                  shouldSkipCoordinateRecord(timeSampleInSec: skipInterval , current: current, lastRecorded: downStreamLast))
            }
            .map { location in
                return CommonCoordinate(location: location, binHeader: b, userInfo: user, isAutoMode: b.allocationNo == self.inAutoMode)
            }
            .eraseToAnyPublisher()
        }
        .sink(receiveValue: { coordinate in
            let value2 = coordinate.1
            if value2 != nil {
                try? Config.Shared.resultDb?.instanceDb?.write { db in
                    do {
                        try value2.save(db)
                    } catch {
                        debugPrint(error)
                    }
                }
            }
        })
        .store(in: &bag)
        /*
        //運行中ステータス
        workingBinHeaders
            .map({ bin -> Bool in
                return bin != nil
            })
            .removeDuplicates()
            .map({ w in
                if w {
                    self.locationPublisher.map({
                        BinState(running: true, location: $0)
                    })
                    .prepend(BinState(running: true))
                } else {
                    Just(BinState(running: false))
                }
            })
            .sink(receiveValue: {
                
            })
            .store(in: &bag)
        */
    }
    
    @objc func methodOfReceivedNotification(notification: Notification) {

        switch notification.name.rawValue {
        case BroadcastEnum.Location.rawValue:
            accept()
            onLocationResult()
        default:
            break
        }
    }
    
    func removeObserver() {
        NotificationCenter.default.removeObserver(self)
    }
    
    private var bin: [BinHeader]? = nil
    private var recordProcessor: BinState? = nil
    
    private var runningBinHeaders: [BinHeaderAndStatus]? {
        return Current.Shared.userRepository?.binHeaderRepo.list()
    }
    /*@Published private var workingBinHeaders =
    Just(Current.Shared.userRepository?.binHeaderRepo.list()?
        .compactMap({ $0 })
        .filter({ item in
            item.status.getStatus() == .Working
        })
        .compactMap({ it -> BinHeader in
            return it.header
        }) ?? []
    )
    .shareReplay(1)*/
    
    private func accept() {
        if let _ = Current.Shared.loggedUser?.userInfo {
            
            if let l = runningBinHeaders {
                var hasItem = false
                for item in l {
                    if item.header.allocationNo == inAutoMode {
                        hasItem = true
                    }
                }
                
                if !hasItem {
                    cancelAutoMode()
                }
                
                var tempBin: [BinHeader] = []
                for item in l {
                    if item.status.getStatus() == .Working {
                        tempBin.append(item.header)
                    }
                }
                
                bin = tempBin
            }
        }
    }
    
    private func tryConnect() {
        if bin == nil || bin!.isEmpty {
            recordProcessor = nil
        }
    }
    
    private func onLocationResult() {
        if let _ = Current.Shared.loggedUser?.userInfo {
            if let c = Current.Shared.lastLocation {
                locationPublisher.send(c)
                if let b = bin {
                    recordProcessor = BinState(running: true, location: c)
                    
                    for header in b {
                        let k = header.allocationNo
                        
                        if k == inAutoMode {
                            autoModeProcess(allocationNo: k, location: c)
                        }
                    }
                }
            }
        }
    }
    
    class M {
        var binDetailAndStatus: BinDetailAndStatus
        var check: (Int, Bool)?
        init(binDetailAndStatus: BinDetailAndStatus, check: (Int, Bool)?){
            self.binDetailAndStatus = binDetailAndStatus
            self.check = check
        }
        
        var detail: BinDetail {
            return binDetailAndStatus.detail
        }
        
        var status: WorkStatus {
            return binDetailAndStatus.status
        }
        
        var inRange: Bool? {
            return check?.1
        }
    }
    
    @Published private var autoStart: [M]? = nil
    private func autoModeProcess(allocationNo: String, location: CLLocation) {
        let detailRepo = Current.Shared.userRepository?.binDetailRepo
        let workRepo = Current.Shared.userRepository?.workRepo

        let detailList = detailRepo!.binDetailWithStatusList(allocationNo: allocationNo)!
            .map { M(binDetailAndStatus: $0, check: $0.detail.checkDeliveryDeviation(compare: location))}
        
        let group = Dictionary.init(grouping: detailList, by: { $0.status.getStatus() })
        if let working = group[WorkStatusEnum.Working]?.first {
            let inRange = working.inRange
            if inRange != false {
                return
            }
            else {
                workRepo?.endWork(binDetail: working.detail, endLocation: location, moveNext: true, isAutoMode: true)
                Helper.Shared.sendBroadcastNotification(.Bin)
                Current.Shared.syncBin() {(isOk, Error) in}
            }
        } else {
            let moving = group[WorkStatusEnum.Moving]?.first(where: { $0.inRange ?? false })
            if moving == nil {
                workRepo?.autoMoveBinDetail(allocationNo: allocationNo)
            }
        }
        
        autoStart = detailList
    }
    
    private func getServiceOrder(binDetailAndStatus: BinDetailAndStatus) -> Int {
        let s = binDetailAndStatus.detail.serviceOrder
        if (s == nil) {
            return Int.max
        } else {
            switch binDetailAndStatus.status.getStatus() {
            case .Finished:
                return ~s!
            default:
                return s!
            }
        }
    }
    
    private func getWorkStatusOrder(wStatus: WorkStatusEnum) -> Int {
        switch wStatus {
        case .Moving:
            return 1
        case .Ready:
            return 2
        case .Finished:
            return 3
        default:
            return 4
        }
    }

    func throttleLatestRecord() -> BinState? {
        return recordProcessor
    }
    
    private var userLocationRequest: Int {
        return Current.Shared.loggedUser?.userInfo.positionGetInterval ?? 5
    }
    private var autoModeLocationRequest = 1
    
    private func ensureLocationRequestMode() {
        Current.Shared.locationHelper?.setLocationRequest(second: inAutoMode.isEmpty ? userLocationRequest : autoModeLocationRequest)
    }
    
    private var inAutoMode: String = Resources.strEmpty
    
    var inAutoModeBin: String {
        return inAutoMode
    }
    
    func setInAutoMode(binHeader: BinHeader) {
        inAutoMode = binHeader.allocationNo
        ensureLocationRequestMode()
        Current.Shared.locationHelper?.onLocationCallBack(nil)
    }
    
    func cancelAutoMode() {
        inAutoMode = Resources.strEmpty
        autoStart = nil
        ensureLocationRequestMode()
    }
    
    private var sensorDetectorDelayStart: SensorDetectorDelayStart? = nil
    private var sensorManager: CMMotionManager = CMMotionManager()
    private class SensorDetectorDelayStart : DelayStart {
        
        private var sensorManager : CMMotionManager
        private var callBack: (Double) -> Void
        private var disposable: Cancellable? = nil
        private var lastTime: Double = 0
        private let limitTime: Int = 10_000
        
        init(sensorManager: CMMotionManager, callBack: @escaping (Double) -> Void) {
            self.sensorManager = sensorManager
            self.callBack = callBack
        }
        
        override func onStart() {
            disposable = sensorManager
                .rxAccelerometerDetectionMillis(sampling: 100, size: 10, limit: self.limitTime, threshold: 4.3)
                .subscribe(on: DispatchQueue.global())
                .sink(receiveValue: { ctime in
                    let dst = ctime - self.lastTime
                    if dst > Double(self.limitTime/1000) {
                        self.lastTime = ctime
                        self.callBack(ctime)
                    }
                })
        }
        
        override func onStop(isStarted: Bool) {
            disposable?.cancel()
            sensorManager.stopDeviceMotionUpdates()
            if (isStarted) {
                disposable = nil
            }
        }
    }
}

File: LogAndroid/TrustarApp/src/main/kotlin/jp/co/toukei/log/trustar/user/BinLocationTask.kt
package jp.co.toukei.log.trustar.user

import android.location.Location
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.Priority
import io.reactivex.rxjava3.core.Flowable
import io.reactivex.rxjava3.disposables.Disposable
import io.reactivex.rxjava3.internal.disposables.ListCompositeDisposable
import io.reactivex.rxjava3.processors.BehaviorProcessor
import io.reactivex.rxjava3.processors.PublishProcessor
import jp.co.toukei.log.lib.addTo
import jp.co.toukei.log.lib.distinctUntilOptionalChanged
import jp.co.toukei.log.lib.observeOnIO
import jp.co.toukei.log.lib.optional
import jp.co.toukei.log.lib.optionalSwitchMap
import jp.co.toukei.log.lib.orElseNull
import jp.co.toukei.log.lib.replayThenAutoConnect
import jp.co.toukei.log.lib.subscribeOnIO
import jp.co.toukei.log.lib.util.LocationRequestHelper
import jp.co.toukei.log.lib.util.filterWithLast
import jp.co.toukei.log.trustar.App
import jp.co.toukei.log.trustar.BroadcastEvent
import jp.co.toukei.log.trustar.BuildConfig
import jp.co.toukei.log.trustar.Config
import jp.co.toukei.log.trustar.Current
import jp.co.toukei.log.trustar.SplitBuildVariant
import jp.co.toukei.log.trustar.compose.ComposeData
import jp.co.toukei.log.trustar.db.result.entity.CommonCoordinate
import jp.co.toukei.log.trustar.db.result.entity.CommonRest
import jp.co.toukei.log.trustar.db.result.entity.CommonSensorCsv
import jp.co.toukei.log.trustar.db.user.entity.BinDetail
import jp.co.toukei.log.trustar.detect.DetectRestEvent
import jp.co.toukei.log.trustar.detect.SensorRecord
import jp.co.toukei.log.trustar.detect.eachWorkingBin
import jp.co.toukei.log.trustar.detect.restEvent
import jp.co.toukei.log.trustar.detect.sensorDetect
import jp.co.toukei.log.trustar.enum.WorkStatusEnum
import jp.co.toukei.log.trustar.isWorking
import jp.co.toukei.log.trustar.repo.BinDetailRepo
import jp.co.toukei.log.trustar.repo.BinHeaderRepo
import jp.co.toukei.log.trustar.repo.WorkRepo
import jp.co.toukei.log.trustar.shouldSkipCoordinateRecord
import jp.co.toukei.log.trustar.shouldSkipLocation
import kotlinx.coroutines.rx3.asFlowable
import java.util.Optional
import java.util.concurrent.TimeUnit
import kotlin.math.abs

/**
 * GPSデータ取得　
 */
class BinLocationTask(
    loggedUser: LoggedUser,
) : Disposable {

    class BinState(
        @JvmField val running: Boolean,
        @JvmField val location: Location? = null,
    )

    private val locationPublisher = PublishProcessor.create<Location>()

    private class M(
        @JvmField val detail: BinDetail,
        @JvmField val inRange: Boolean?,
    )

    private class MM(
        val allocationNo: String,
        val location: Location,
        val ls: List<M>,
    )

    private val autoStart = PublishProcessor.create<List<M>>()

    private val helper: LocationRequestHelper = LocationRequestHelper(
        loggedUser.locationHelper,
        { c ->
            Current.lastLocation = c
            locationPublisher.onNext(c)
        },
        { e ->
            App.gmsLocationErrFlow.tryEmit(e.cause ?: e)
        }
    )

    private val binStateProcessor: BehaviorProcessor<BinState> = BehaviorProcessor.create()

    fun throttleLatestRecord(timeoutSec: Long): Flowable<BinState> {
        return binStateProcessor.throttleLatest(timeoutSec, TimeUnit.SECONDS, true)
            .onBackpressureLatest()
            .observeOnIO(1)
    }


    private val userLocationRequest =
        LocationRequest.Builder(loggedUser.userInfo.getIntervalInSec * 1000L)
            .setPriority(Priority.PRIORITY_HIGH_ACCURACY)
            .setMinUpdateIntervalMillis(1000L)
            .build()

    private val autoModeLocationRequest = LocationRequest.Builder(1000L)
        .setPriority(Priority.PRIORITY_HIGH_ACCURACY)
        .setMinUpdateIntervalMillis(1000L)
        .build()

    private val disposables = ListCompositeDisposable()

    private val headerRepo = BinHeaderRepo(loggedUser.userDB)
    private val detailRepo = BinDetailRepo(loggedUser.userDB)
    private val workRepo = WorkRepo(loggedUser.userDB)

    private val workingBinHeaders = headerRepo.list()
        .map { all ->
            all.mapNotNull { if (it.status.isWorking()) it.header else null }
        }
        .observeOnIO()
        .subscribeOnIO()
        .replayThenAutoConnect(disposables)


    init {
        autoStart.onBackpressureLatest()
            .observeOnIO(1)
            .scan(Optional.empty<BinDetail>()) { acc, e ->
                val inRange: List<M> = e.filter {
                    it.inRange == true && it.detail.workStatus.isNotWorking()
                }
                val b = acc.orElseNull()?.let { b -> inRange.firstOrNull { it.detail.sameKey(b) } }
                    ?: inRange.minWithOrNull { o1, o2 ->
                        compareValuesBy(o1, o2,
                            {
                                when (it.detail.workStatus) {
                                    WorkStatusEnum.Moving -> 1
                                    WorkStatusEnum.New -> 2
                                    WorkStatusEnum.Finished -> 3
                                    else -> 4
                                }
                            },
                            {
                                val s = it.detail.serviceOrder
                                if (s == null) Int.MAX_VALUE
                                else when (it.detail.workStatus) {
                                    WorkStatusEnum.Finished -> s.inv()
                                    else -> s
                                }
                            }
                        )
                    }
                Optional.ofNullable(b?.detail)
            }
            .distinctUntilChanged { t1, t2 ->
                val b1 = t1.orElseNull()
                val b2 = t2.orElseNull()
                b1 != null && b2 != null && b1.sameKey(b2)
            }
            .switchMap {
                val detail = it.orElseNull()
                if (detail == null) {
                    Flowable.empty()
                } else {
                    val delayInMin = BinDetail.stayTimeDelayInMin(detail).toLong()
                    Flowable.just(detail).delay(delayInMin, TimeUnit.MINUTES)
                }
            }
            .onBackpressureDrop()
            .observeOnIO(1)
            .retry()
            .subscribe {
                val w = it.work
                if (w != null) {
                    workRepo.startWork(
                        allocationNo = it.allocationNo,
                        allocationRowNo = it.allocationRowNo,
                        work = ComposeData.Work(w.workCd, w.workNm.orEmpty(), 0),
                        fromLocation = Current.lastLocation,
                        addIfFinished = true,
                        isAutoMode = true
                    )
                }
            }
            .addTo(disposables)
    }

    override fun dispose() {
        disposables.dispose()
        helper.disconnect(true)
    }

    override fun isDisposed(): Boolean {
        return disposables.isDisposed
    }

    private val inAutoModeAllocationNo = BehaviorProcessor.create<Optional<String>>()

    @JvmField
    val inAutoModeBin: Flowable<Optional<String>> = inAutoModeAllocationNo

    fun setInAutoMode(allocationNo: String) {
        inAutoModeAllocationNo.onNext(allocationNo.optional())
    }

    fun cancelAutoMode() {
        inAutoModeAllocationNo.onNext(Optional.empty())
    }

    init {
        val user = loggedUser.userInfo

        //休憩判定
        restEvent(workingBinHeaders, locationPublisher, user)
            .observeOnIO()
            .subscribe { (allocationNo, r) ->
                if (r is DetectRestEvent.MoveDetected) {
                    val c = CommonRest(
                        userInfo = user,
                        allocationNo = allocationNo,
                        startLocation = r.rest.startRestLocation,
                        endDate = r.restEndTimeBasedOsDate(),
                        elapsed = r.elapsedRestMillis(),
                    )
                    loggedUser.resultDb.commonRestDao().insert(c)
                    Current.syncRest()
                }
                SplitBuildVariant.debugRestEvent(allocationNo, r)
            }
            .addTo(disposables)

        val d = user.decimationRange
        val dao = loggedUser.userDB.sensorDetectEventDao()
        val dao2 = Config.resultDb.commonSensorCsvDao()
        // todo. non-working bin changes

        //センサー検知
        sensorDetect(
            workingBinHeaders,
            binHeaderDelayInMillis = user.detectStartTimingInMin * 60_000L,
            workDelayInMillis = 5000,
        ) { detailRepo.countWorkingStatus(it).map { c -> c == 0 } }
            .observeOnIO()
            .subscribe {
                when (it) {
                    is SensorRecord.DropRecent -> {
                        val no = it.allocationNo
                        dao.deleteByTimeRange(no, it.from, it.to)
                        if (it.isFinished) {
                            val ls = dao.findBy(no)
                            if (ls.isNotEmpty()) {
                                Config.dateFormatterSensorCsv.use { formatter ->
                                    val csv = CommonSensorCsv(
                                        user, no,
                                        ls.joinToString("\n") {
                                            it.csvRow(formatter)
                                        }.toByteArray()
                                    )
                                    dao2.insert(csv)
                                    dao.deleteBy(no)
                                }
                                Current.syncSensorCsvUpload()
                            }
                        }
                    }

                    is SensorRecord.Record -> {
                        val n = it.event
                        val l = dao.latestEvent(n.allocationNo)
                            ?.eventRecord
                            ?.distanceTo(n.eventRecord)
                        if (l == null || d == null || l > d) {
                            dao.insert(n)
                        } else {
                            if (BuildConfig.isDebug) {
                                SplitBuildVariant.debugSensorEvent(it, true)
                                return@subscribe
                            }
                        }
                    }
                }

                if (BuildConfig.isDebug) {
                    SplitBuildVariant.debugSensorEvent(it, false)
                }
            }
            .addTo(disposables)

        //センサー検知レコード位置更新
        locationPublisher
            .observeOnIO()
            .subscribe { location ->
                dao.latestEventEachGroup()
                    .mapNotNull { sde ->
                        val prevLocationTime = sde.locationTimestamp ?: 0L
                        val currLocationTime = location.time
                        val eventTime = sde.eventRecord.date
                        val td1 = abs(eventTime - prevLocationTime)
                        val td2 = abs(eventTime - currLocationTime)
                        if (td1 > td2) {
                            sde.copyOf(location)
                        } else null
                    }
                    .let {
                        dao.insertOrReplace(it)
                    }
            }
            .addTo(disposables)

        val skipRange = user.decimationRange ?: 50
        val skipInterval = user.getIntervalInSec

        //位置レコード
        workingBinHeaders
            .eachWorkingBin { b ->
                locationPublisher
                    .observeOnIO()
                    .filterWithLast { upStreamLast, downStreamLast, current ->
                        !(shouldSkipLocation(skipRange, current, upStreamLast, downStreamLast) ||
                                shouldSkipCoordinateRecord(skipInterval, current, downStreamLast))
                    }
                    .map { location ->
                        val a = inAutoModeAllocationNo.value?.orElseNull()
                        val autoMode = b.allocationNo == a
                        CommonCoordinate(location, b, user, autoMode)
                    }
                    .onErrorComplete()
            }
            .subscribeOnIO()
            .subscribe { coordinate ->
                Config.resultDb.commonCoordinateDao().insertOrReplace(coordinate)
            }
            .addTo(disposables)

        //運行中ステータス
        workingBinHeaders
            .map { it.isNotEmpty() }
            .distinctUntilChanged()
            .switchMap { w ->
                if (w) {
                    locationPublisher
                        .map { BinState(true, it) }
                        .startWithItem(BinState(true))
                } else {
                    Flowable.just(BinState(false))
                }
            }
            .subscribe(binStateProcessor::onNext)
            .addTo(disposables)

        //自動モードオフ
        inAutoModeAllocationNo
            .optionalSwitchMap { no ->
                workingBinHeaders.map { w ->
                    w.none { it.allocationNo == no }
                }
            }
            .subscribe {
                if (it) {
                    cancelAutoMode()
                }
            }
            .addTo(disposables)

        //自動モード作業
        Flowable
            .combineLatest(
                inAutoModeAllocationNo,
                workingBinHeaders
            ) { a, w ->
                val no = a.orElseNull()
                if (no == null || w.none { it.allocationNo == no }) Optional.empty() else a
            }
            .distinctUntilOptionalChanged()
            .switchMap {
                val allocationNo = it.orElseNull()
                if (allocationNo == null) {
                    Flowable.just(Optional.empty())
                } else {
                    Flowable.combineLatest(
                        detailRepo.binDetailWithStatusList(allocationNo),
                        locationPublisher,
                    ) { list, loc ->
                        MM(
                            allocationNo,
                            loc,
                            list.map {
                                M(
                                    it.detail,
                                    BinDetail.checkDeliveryDeviation(it.detail, loc)?.second
                                )
                            }
                        ).optional()
                    }
                }
            }
            .subscribe { ooo ->
                val oo = ooo.orElseNull()
                if (oo == null) {
                    autoStart.onNext(emptyList())
                } else {
                    val l = oo.ls
                    val working = l.firstOrNull { it.detail.workStatus.isWorking() }
                    if (working != null) {
                        if (working.inRange == false) {
                            val x = working.detail
                            //out of range
                            workRepo.endWork(
                                x.allocationNo,
                                x.allocationRowNo,
                                oo.location,
                                true,
                                isAutoMode = true
                            )
                            autoStart.onNext(l)
                        }
                    } else {
                        val moving = l.firstOrNull {
                            it.detail.workStatus.isMoving() && (it.inRange ?: false)
                        }
                        if (moving == null) {
                            workRepo.autoMoveBinDetail(oo.allocationNo)
                        }
                        autoStart.onNext(l)
                    }
                }
            }
            .addTo(disposables)

        //手動・自動モード位置取得
        Flowable
            .combineLatest(
                workingBinHeaders.map { it.isNotEmpty() },
                inAutoModeAllocationNo.startWithItem(Optional.empty())
            ) { w, a ->
                if (w) {
                    Optional.ofNullable(
                        if (a.orElseNull() == null) userLocationRequest
                        else autoModeLocationRequest
                    )
                } else {
                    Optional.empty()
                }
            }
            .throttleLatest(1, TimeUnit.SECONDS, true)
            .distinctUntilOptionalChanged()
            .subscribe {
                val r = it.orElseNull()
                if (r != null) {
                    helper.setRequest(r)
                } else {
                    helper.disconnect(true)
                }
            }
            .addTo(disposables)

        //位置取得条件エラー場合
        helper.latestErr
            .switchMap {
                val e = it.orElseNull()
                if (e == null) Flowable.empty()
                else Flowable.just(e).delay(5, TimeUnit.SECONDS).repeat()
            }
            .onBackpressureLatest()
            .observeOnIO()
            .subscribe {
                //todo ???
                helper.reconnected(true)
            }
            .addTo(disposables)
        App.broadcastFlow()
            .asFlowable()
            .subscribe {
                if (it is BroadcastEvent.LocationSettingsChanged)
                    helper.reconnected(true)
            }
            .addTo(disposables)
    }
}
